homeassistant/components/http/forwarded.py

```diff
@@ -0,0 +1,135 @@
+"""Middleware to handle forwarded data by a reverse proxy."""
+from ipaddress import ip_address
+import logging
+
+from aiohttp.hdrs import X_FORWARDED_FOR, X_FORWARDED_HOST, X_FORWARDED_PROTO
+from aiohttp.web import HTTPBadRequest, middleware
+
+from homeassistant.core import callback
+
+_LOGGER = logging.getLogger(__name__)
+
+# mypy: allow-untyped-defs
+
+
+@callback
+def setup_forwarded(app, trusted_proxies):
+    """Create forwarded middleware for the app.
+
+    Process IP addresses, proto and host information in the forwarded for headers.
+
+    `X-Forwarded-For: client, proxy1, proxy2`
+
+    We go trough the list from the right side, and skip all entries that are in our
+    trusted proxies list. The first non-trusted IP is used as the client IP. If all
+    items in the X-Forwarded-For are trusted, including the most left item (client),
+    the most left item is used.
+
+    `X-Forwarded-Proto: client, proxy1, proxy2`
+    e.g, `X-Forwarded-Proto: https, http, http`
+    OR `X-Forwarded-Proto: https` (one entry, even with multiple proxies)
+
+    The X-Forwarded-Proto is determined based on the corresponding entry of the
+    X-Forwarded-For header that is used / chosen as the client IP. However,
+    some proxies, for example, Kubernetes NGINX ingress, only retain one element
+    in the X-Forwarded-Proto header. In that case, we'll just use what we have.
+
+    `X-Forwarded-Host: example.com`
+
+    If the previous headers are processed successfully, and the X-Forwarded-Host is
+    present, it will be used.
+
+    Additionally:
+      - If no X-Forwarded-For header is found, processing of all headers is skipped.
+      - If multiple instances of a X-Forwarded-For, X-Forwarded-Proto or
+        X-Forwarded-Host are found, a HTTP 400 status code is thrown.
+      - If malformed or invalid (IP) data in X-Forwarded-For header is found,
+        a HTTP 400 status code is thrown.
+      - The connected client peer on the socket of the incoming connection,
+        must be trusted for any processing to take place.
+    """
+    #
+    @middleware
+    async def forwarded_middleware(request, handler):
+        """Process forwarded data by a reverse proxy."""
+        overrides = {}
+
+        # Handle X-Forwarded-For
+        forwarded_for = request.headers.getall(X_FORWARDED_FOR, [])
+        if not forwarded_for:
+            # No forwarding headers, continue as normal
+            return await handler(request)
+
+        # Ensure the IP of the connected peer is trusted
+        connected_ip = ip_address(request.transport.get_extra_info("peername")[0])
+        if not any(connected_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+            # Not trusted, continue as normal
```

Comment:
Should we log these cases? As in, we have an `X-Forwarded-For` header, but the directly connected peer is untrusted.

Reply:
I guess the answer is yes. This will actually help users finding out why the remote IP doesn't work as expected.

Reply:
Added in 3071fce